pkg/ldbwriter/ldb_callback_writer.go

```diff
@@ -3,33 +3,105 @@ package ldbwriter
 import (
 	"context"
 	"database/sql"
-
 	"github.com/segmentio/ctlstore/pkg/schema"
 	"github.com/segmentio/ctlstore/pkg/sqlite"
 	"github.com/segmentio/events/v2"
+	"github.com/segmentio/stats/v4"
 )
 
 // CallbackWriter is an LDBWriter that delegates to another
 // writer and then, upon a successful write, executes N callbacks.
 type CallbackWriter struct {
-	DB           *sql.DB
-	Delegate     LDBWriter
-	Callbacks    []LDBWriteCallback
+	DB        *sql.DB
+	Delegate  LDBWriter
+	Callbacks []LDBWriteCallback
+	// Buffer between SQLite preupdate Hook and this code
 	ChangeBuffer *sqlite.SQLChangeBuffer
+	// Accumulated changes across multiple ApplyDMLStatement calls
+	transactionChanges []sqlite.SQLiteWatchChange
+}
+
+func (w *CallbackWriter) inTransaction() bool {
+	return w.transactionChanges != nil
+}
+
+func (w *CallbackWriter) beginTransaction(ledgerSequence schema.DMLSequence) {
+	if len(w.transactionChanges) > 0 {
+		// This should never happen, but just in case...
+		stats.Add("ldb_changes_abandoned", len(w.transactionChanges))
+		events.Log("error: abandoned %{count}d changes from incomplete transaction, current statement's ledger sequence: %{sequence}d",
+			len(w.transactionChanges), ledgerSequence)
+	}
+	w.transactionChanges = make([]sqlite.SQLiteWatchChange, 0)
+	// TODO: Figure out if we wanna use a gauge or a counter here
+	stats.Set("ldb_changes_accumulated", 0)
```

Comment:
what are we trying to observe with this metric? The number of transactions handled? The size of a transaction? The fact that the reflector was processing a transaction at a certain point?

Reply:
Trying to get a sense of how big the changelog was getting to before we flush it out.  I'm inexperienced in crafting these kinds of metrics, so very welcoming of suggestions here!

Reply:
In that case, I'd go with `.Observe()`, as it will produce percentiles as well as the max

Reply:
Per offline discussion, just ditching these active-accumulation metrics for now, in favor of the `ldb_changes_written` metric.